fluent/migrate/transforms.py

```diff
@@ -92,42 +85,106 @@ def __call__(self, ctx):
-            if elem_type is FTL.TextElement:
-                # Join adjacent TextElements.
-                text = FTL.TextElement(''.join(elem.value for elem in elems))
-                # And remove empty ones.
-                if len(text.value) > 0:
-                    pruned.append(text)
+    def normalize_text_content(elements):
+        '''Normalize PatternElements with text content.
+
+        Convert TextElements and StringExpressions into TextElements and join
+        adjacent ones.
+        '''
+
+        def get_text(element):
+            if isinstance(element, FTL.TextElement):
+                return element.value
+            elif isinstance(element, FTL.Placeable):
+                if isinstance(element.expression, FTL.StringExpression):
+                    return element.expression.value
+
+        joined = []
+        for current in elements:
+            current_text = get_text(current)
+            if current_text is None:
+                joined.append(current)
+                continue
+
+            previous = joined[-1] if len(joined) else None
+            if isinstance(previous, FTL.TextElement):
+                previous.value += current_text
+            elif len(current_text) > 0:
+                # Normalize to a TextElement
+                joined.append(FTL.TextElement(current_text))
+        return joined
+
+    @staticmethod
+    def preserve_whitespace(elements):
+        # Handle empty values
+        if len(elements) == 0:
+            return [FTL.Placeable(FTL.StringExpression(''))]
+
+        re_leading = re.compile(r'^(?P<whitespace>\s+)(?P<text>.*?)$')
+        re_trailing = re.compile(r'^(?P<text>.*?)(?P<whitespace>\s+)$')
+
+        def extract_whitespace(regex, element):
+            '''Extract leading or trailing whitespace from a TextElement.
+
+            Return a tuple of (Placeable, TextElement) in which the Placeable
+            encodes the extracted whitespace as a StringExpression and the
+            TextElement has the same amount of whitespace removed. The
+            Placeable with the extracted whitespace is always returned first.
+            '''
+            match = re.search(regex, element.value)
+            if match:
+                whitespace = match.group('whitespace')
+                empty_expr = FTL.Placeable(FTL.StringExpression(whitespace))
+                if whitespace == element.value:
+                    return empty_expr, None
+                else:
+                    return empty_expr, FTL.TextElement(match.group('text'))
             else:
-                pruned.extend(elems)
-        return pruned
+                return None, element
+
+        if isinstance(elements[0], FTL.TextElement):
+            ws, text = extract_whitespace(re_leading, elements[0])
+            elements[:1] = [ws, text]
+
+        if isinstance(elements[-1], FTL.TextElement):
+            ws, text = extract_whitespace(re_trailing, elements[-1])
+            elements[-1:] = [text, ws]
```

Comment:
... and here ...

Reply:
Always returning a 2-tuple makes it easy to switch the order in this particular case here. Note that `extract_whitespace` doesn't know if it's processing the leading or the trailing whitespace.